Rewrite HTTP API documentation (#1761)
replicate/cog: docs/http.md

# HTTP API

> [!TIP]
> For information about how to run the HTTP server, 
> see [our documentation to deploying models](deploy.md).

When you run a Docker image built by Cog, 

> it serves an HTTP API for making predictions.

Reviewer: This would be a larger change (and therefore perhaps best made in a separate PR?) but we've been talking about moving away from using the term "prediction" in favor of more software-engineering-y terms like "run" (context: [style guide](https://www.notion.so/replicate/Style-guide-377cfdbd142540a9b3177734b63d0baf?pvs=4#8a0bfae63f514b2d955cf332cf429cfb)).

So this would become something like "it serves an HTTP API for running the model" ? Again, mostly just bringing it up here for discussion; this doc has so many mentions of prediction that I think if we do want to change the language we'd want to do that in a followup.

Author: Yes, please! Caveat:

You run models, models are run — I agree with that. When you run the model for inference, the method called is `predict` and the record / result of your running the model is a "prediction". I don't think we should start calling that a "run" (if only because of its scatalogical plural form, but also because that's a big API change).

The server supports both synchronous and asynchronous prediction creation:

- **Synchronous**: 
  The server waits until the prediction is completed
  and responds with the result.
- **Asynchronous**: 
  The server immediately returns a response 
  and processes the prediction in the background. 

The client can create a prediction asynchronously
by setting the `Prefer: respond-async` header in their request.
When provided, the server responds immediately after starting the prediction 
with `202 Accepted` status and a prediction object in status `processing`.

> [!NOTE]
> The only supported way to receive updates on the status of predictions
> started asynchronously is using [webhooks](#webhooks). 
> Polling for prediction status is not currently supported.

> The server also provides a way to create predictions idempotently,

Reviewer: I want to make this a bit more user-active, but without losing any technical nuance. What do you think of this?

```suggestion
You can also use certain server endpoints to create predictions idempotently,
```

while the prediction is still running, 
no new prediction is created. 
Instead, the client receives a `202 Accepted` response
with the initial state of the prediction.

---

Here's a summary of the prediction creation endpoints:

| Endpoint                           | Header                  | Behavior                     |
| ---------------------------------- | ----------------------- | ---------------------------- |
| `POST /predictions`                | -                       | Synchronous, non-idempotent  |
| `POST /predictions`                | `Prefer: respond-async` | Asynchronous, non-idempotent |
| `PUT /predictions/<prediction_id>` | -                       | Synchronous, idempotent      |
| `PUT /predictions/<prediction_id>` | `Prefer: respond-async` | Asynchronous, idempotent     |

Choose the endpoint that best fits your needs:

- Use synchronous endpoints when you want to wait for the prediction result.
- Use asynchronous endpoints when you want to start a prediction 
  and receive updates via webhooks.
- Use idempotent endpoints when you need to safely retry requests 
  without creating duplicate predictions.

## Webhooks

> The client can provide a `webhook` parameter in the request body 

Reviewer: ```suggestion
You can provide a `webhook` parameter in the client request body
```

Making it a bit more active
